modules/dhcp_common/free_ips.rb

```diff
@@ -0,0 +1,157 @@
+    def mark_ip_as_allocated(ip_address)
+      @m.synchronize do
+        @allocated_ips << ip_address
+        @allocation_timestamps.push([ip_address, time_now + DEFAULT_ADDRESS_KEEP_TIME])
+      end
+    end
+
+    def clean_up_allocated_ips
+      @m.synchronize do
+        loop do
+          logger.debug("Starting allocated ip addresses cleanup pass...")
+          break if @allocation_timestamps.first.nil? || @allocation_timestamps.first[1] > time_now
+          ip, _ = @allocation_timestamps.shift
+          @allocated_ips.delete(ip)
+          logger.debug("#{ip} marked as free.")
+        end
+      end
+    end
+
+    def start
+      logger.info("Starting allocated ip address maintenance (used by unused_ip call).")
+      @timer_task = Concurrent::TimerTask.new(:execution_interval => DEFAULT_CLEANUP_INTERVAL) { clean_up_allocated_ips }
+      @timer_task.execute
+    end
+
+    def stop
+      @timer_task.shutdown unless @timer_task.nil?
+    end
+
+    def time_now
+      Time.now.to_i
+    end
+
+    def find_free_ip(from_address, to_address, records)
+      start_address_i, num_of_addresses = address_range_with_start_and_end(from_address, to_address)
+
+      record_ips = Set.new(records.collect{|record| record.ip})
+      random_index(num_of_addresses + 1) do |i|
+        possible_ip = ::Proxy::DHCP.i_to_ipv4(start_address_i + i)
+        next if @m.synchronize do
+          if @allocated_ips.include?(possible_ip) || record_ips.include?(possible_ip)
+            true
+          else
+            mark_ip_as_allocated(possible_ip)
+            false
+          end
+        end
+        begin
+          logger.debug "Searching for free IP - pinging #{possible_ip}."
+          if tcp_pingable?(possible_ip) || icmp_pingable?(possible_ip)
+            logger.debug "Found a pingable IP(#{possible_ip}) address which does not have a Proxy::DHCP record."
+          else
+            logger.debug "Found free IP #{possible_ip} out of a total of #{num_of_addresses} free IPs."
+            return possible_ip
+          end
+        rescue Exception => e
+          logger.error("Exception when pinging #{possible_ip}. Skipping the address.", e)
+        end
+      end
+
+      logger.warn "No free IPs in range #{from_address}..#{to_address}."
+      nil
+    end
+
+    def address_range_with_start_and_end(start_address, end_address)
+      as_i = Proxy::DHCP.ipv4_to_i(start_address)
+      [as_i, Proxy::DHCP.ipv4_to_i(end_address) - as_i]
+    end
+
+    # returns indexes for an array in random order
+    def random_index(total_number_of_indices)
+      raise ArgumentError.new("Zero or negative number of indices") if total_number_of_indices <= 0
+      rng = Random.new(Time.now.to_i)
+      past_indices = Set.new
+      loop do
+        break if past_indices.size >= total_number_of_indices
+        current_index = rng.rand(total_number_of_indices)
+        next if past_indices.add?(current_index).nil?
+        yield(current_index)
+      end
```

Comment:
Sidenote for reviewers: when reaching it's full capacity, bigger ranges could possibly spend a lot of CPU cycles in this loop. But this is ok, we are theoretically swapping memory efficiency for CPU efficiency (in practice this code is much better because creating huge array was basically a blocker).

Also it is worth nothing we are using PSEUDO RANDOM GENERATOR which is actually better than real random generator. PRNG produces numbers in cycles making sure the whole space is walked in sequence. In Ruby the Mersenne Twister algorithm is used. One nitpick - would be great to ensure the generator is really pseudo, so there won't be regression in future Ruby version. We could achieve this with a simple test going through whole space of let's say hundred of random numbers. Second, pseudo random generators won't be blocked when OS don't have enough entropy, which is what we want to avoid.